### ISSUE TITLE
[RETRACE] Could the texture read back type the inferred with glGetInternalFormat?

### ISSUE THREAD

**Issue by svenpilz** [apitrace/apitrace#344]
I was running into #343, the stored image was over 3GB in size, while it actually was only about 900MB.

The reason seems to be the guessing of the `read back format`. I was debugging an application that was using an `GL_RGBA` `GL_UNSIGNED_BYTE` texture array like the following

```
glTexImage3D(target = GL_TEXTURE_2D_ARRAY, level = 0, internalformat = GL_RGBA,
width = 7360, height = 4912, depth = 6, border = 0, format = GL_RGBA,
type = GL_UNSIGNED_BYTE, pixels = ...)
```

So it was using unsized types and this is where the heuristic of `internalFormatDescs` comes into play. To cut the long story short :-) : Couldn't `glGetInternalFormat` be used to get the actual type?

**Comment by jrfonseca** [comment]
Right. `glretrace` doesn't know how precise an unsized internalformat like `GL_RGBA` is, so it ends up using floats.

> Couldn't glGetInternalFormat be used to get the actual type?

Yes, per https://www.opengl.org/sdk/docs/man/docbook4/xhtml/glGetInternalformat.xml we could use `GL_INTERNALFORMAT_*_SIZE/TYPE` or `GL_GET_TEXTURE_IMAGE_FORMAT/TYPE` for a better guess on how the implementation is storing it.

Do you want to venture a patch? 

Still, I wonder if I should just assume that unsized internal formats == 8 bits...

**Comment by jrfonseca** [comment]
> Still, I wonder if I should just assume that unsized internal formats == 8 bits...

I implemented this for now. As `glGetInternalformat` is only available on 4.2.

I also added a TODO comment to one day leverage `glGetInternalformat` as you suggest, but for the time being this should be good enough.  We actually have more pressing issues dumping images, like limitates with integer formats, etc.
